### ISSUE TITLE
Intermittent CUDA_ERROR_ILLEGAL_ADDRESS errors during back propagation for some networks when using XLA.

### ISSUE THREAD

**Issue by nryant** [tensorflow/tensorflow#6816]
### Summary
In playing around with the XLA JIT compiler I've experienced intermittent CUDA_ERROR_ILLEGAL_ADDRESS errors during back propagation for some network structures. While other structures may also fail, I can confirm that this happens with simple fully-connected networks when dropout is applied to the hidden units. I have not, however, observed errors when not using XLA or when using XLA, but no hidden dropout.

### Environment info
- Operating System: Ubuntu 14.04 LTS
- GPU: Titan X
- NVIDIA driver: 375.26
- CUDA: 8.0
- cuDNN: 5.1
- see also: [cuda_libs.txt](https://github.com/tensorflow/tensorflow/files/703021/cuda_libs.txt)
-  bazel: 4.3
- TensorFlow: f8fcd85b8151d287ad09a9323fabf1c5025775b7
- XLA: Yes
- TensorFlow compile command:

    `bazel build --copt=-march=native -c opt  --config=cuda //tensorflow/tools/pip_package:build_pip_package`
    

### Minimal reproducible example 
See https://gist.github.com/nryant/8cea9bb79d7bdb965167e917d8d5aa8b for the script `test_dropout.py`, which repeatedly builds a small network consisting of a single hidden layer of 512 units and performs 100 fprop/bprop steps. It accepts three command line arguments:

- `--bprop` enables backpropagation (by default, only forward propagation is performed)
- `--xla` enables XLA at the session level
- `--dropout` specifies the dropout proportion

Also see log files produced from running the script with various arguments:

- No XLA, fprop + bprop, and dropout: https://gist.github.com/nryant/83829ada37a9b0c221a76184c5e232fa#file-noxla_fprop_bprop_dropout-log
- XLA and fprop: https://gist.github.com/nryant/83829ada37a9b0c221a76184c5e232fa#file-xla_fprop-log
- XLA, fprop, and dropout: https://gist.github.com/nryant/83829ada37a9b0c221a76184c5e232fa#file-xla_fprop_dropout-log
- XLA, fprop + bprop, and dropout: https://gist.github.com/nryant/83829ada37a9b0c221a76184c5e232fa#file-xla_fprop_bprop_dropout-log

The only combination for which we ever observe an error is XLA, fprop + bprop, and dropout, which pretty reliable blows up in the first few iterations.

**Comment by jart** [comment]
@jlebar Mind taking a look at this compiler/xla issue report?

**Comment by jlebar** [comment]
Of course, I'll see what I can figure out here.  Thank you for the detailed bug report, Neville.

**Comment by jlebar** [comment]
Sorry for the long wait; I had a fire to put out elsewhere.

I was able to reproduce the failure locally at f8fcd85 and at current HEAD, 1084748efa.

Nothing more to report, beyond that I'm looking at it now.

**Comment by jlebar** [comment]
Running this through cuda-memcheck, I'm seeing a bunch of errors of the form

    ========= Invalid __global__ read of size 4
    =========     at 0x000000b8 in _copy
    =========     by thread (13,0,0) in block (866,0,0)
    =========     Address 0x3bdac0c83cce9646 is misaligned

That address is indeed misaligned.

The body of the _copy kernel (which you can get by setting the TF_CPP_MIN_VLOG_LEVEL env var to 2) is

	.visible .entry _copy(
		.param .u64 _copy_param_0,
		.param .u64 _copy_param_1,
		.param .u64 _copy_param_2
	)
	.maxntid 32, 1, 1
	{
		.reg .pred 	%p<2>;
		.reg .b32 	%r<6>;
		.reg .b64 	%rd<10>;

		mov.u32 	%r1, %ctaid.x;
		mov.u32 	%r2, %tid.x;
		shl.b32 	%r3, %r1, 5;
		or.b32  	%r4, %r3, %r2;
		setp.lt.u32 	%p1, %r4, 401408;
		@%p1 bra 	LBB1_2;
		bra.uni 	LBB1_1;
	LBB1_2:
		ld.param.u64 	%rd4, [_copy_param_0];
		ld.param.u64 	%rd5, [_copy_param_1];
		cvta.to.global.u64 	%rd1, %rd5;
		cvta.to.global.u64 	%rd2, %rd4;
		cvt.u64.u32 	%rd3, %r4;
		ld.global.u64 	%rd6, [%rd2+8];
		shl.b64 	%rd7, %rd3, 2;
		add.s64 	%rd8, %rd6, %rd7;
		ld.u32 	%r5, [%rd8];
		add.s64 	%rd9, %rd1, %rd7;
		st.global.u32 	[%rd9], %r5;
	LBB1_1:
		ret;
	}

There's only one 32-bit read in this kernel, `ld.u32 %r5, [%rd8]`, so presumably that's the bad one.

`rd8` is the sum of `%rd3 << 2` and a value read from memory.  Clearly `%rd3 << 2` is aligned, so the problem must be that the value we're reading from memory is not.

Not sure yet where this bad value in memory is coming from.

(Incidentally, we should be able to figure out somehow that rd8 is in the global addr space and emit `ld.global.u32`.  But that's just an efficiency problem, not one of correctness.)

**Comment by jlebar** [comment]
Although _copy has three arguments, we seem to be passing only two when we call it.  (Maybe this is intentional; the kernel never uses the third arg.)  But in addition, those two args are the same pointer.  That...does not seem intentional.

**Comment by jlebar** [comment]
Fixed in 081758b0e5efc1a1591cda068a4866099bf8a3c5.

Thanks for your patience, and thanks again for the high-quality bug report.  (As you can see from the commit, we actually had this in hand six days ago, but for unrelated reasons it took until today before we had a successful uplift to github.)
